### PR TITLE
[NET-784][process-agent] Group connections by similarity

### PR DESCRIPTION
**Pull request by ISauve**
### What does this PR do?

When the `process-agent` receives a list of connections from the `system-probe`, it has to batch these connections into reasonably-sized messages for the backend to process. Currently we batch these connections in the order that we receive them; this change groups similar connections together so that messages seen by the backend are more likely to contain connections with the same remote address IP.

### Motivation

[NET-784](https://datadoghq.atlassian.net/browse/NET-784)

### Additional Notes

If there's fewer than `MaxConnsPerMessage` connections, then we don't bother doing any sorting since all the connections will be in the same batch regardless.

### Describe your test plan

I downloaded 500 sample messages and tested the old batching algorithm vs the new batching algorithm on these messages.
I found that with a `MaxConnsPerMessage` = 600 (the default value), the new batching algorithm resulted in a 75% average decrease in the number of unique rAddrs present in each message. This improvement primarily affected hosts which generated many more connections than `MaxConnsPerMessage`.

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

File: pkg/process/checks/net_test.go [hunk 1]
```diff
@@ -125,8 +116,8 @@ func TestNetworkConnectionBatchingWithDNS(t *testing.T) {
 	for i, c := range chunks {
 		connections := c.(*model.CollectorConnections)
 
-		// Only the first chunk should have a DNS mapping!
-		if i == 0 {
+		// Only the last chunk should have a DNS mapping
```
Comment (mbakht) [inline pkg/process/checks/net_test.go:119]: What is the reason for this change? From the code, it seems only `p[0]` has a remote address with an associated DNS entry.

File: pkg/process/checks/net_test.go [hunk 2]
```diff
@@ -99,22 +99,13 @@ func TestNetworkConnectionBatching(t *testing.T) {
 }
 
 func TestNetworkConnectionBatchingWithDNS(t *testing.T) {
-	p := []*model.Connection{
-		makeConnection(1),
-		makeConnection(2),
-		makeConnection(3),
-		makeConnection(4),
-	}
+	p := makeConnections(4)
```
Comment (mbakht) [inline pkg/process/checks/net_test.go:102]: 👍 Thanks for refactoring common code.

File: pkg/process/checks/net_test.go [hunk 3]
```diff
@@ -145,3 +136,35 @@ func TestNetworkConnectionBatchingWithDNS(t *testing.T) {
...
+		assert.Equal(t, int32(3), connections.GroupSize)
+		assert.Equal(t, 2, len(connections.Connections))
+
+		// make sure the connections with similar remote addresses were grouped together
+		rAddr := connections.Connections[0].Raddr.Ip
+		for _, cc := range connections.Connections {
+			assert.Equal(t, rAddr, cc.Raddr.Ip)
+		}
+	}
+	assert.Equal(t, 6, total)
```
Comment (mbakht) [inline pkg/process/checks/net_test.go:169]: Can we also have a test for the case when remote addresses are same but PID as different?

File: pkg/process/checks/net_test.go (hunk 1 above)
Comment (ISauve) [reply pkg/process/checks/net_test.go:119]: When the connections are batched they are sorted by `rAddr`, and since the other `rAddrs` are simply `""` the connection with `rAddr = "1.1.2.3"` ends up at the end of the list

**Review by leeavital — APPROVED** [review]

File: pkg/process/checks/net.go [hunk 4]
```diff
@@ -171,6 +172,16 @@ func batchConnections(
 
 	dnsEncoder := model.NewV1DNSEncoder()
 
+	if len(cxs) > cfg.MaxConnsPerMessage {
+		// Sort connections by remote IP/PID in order to be friendlier to network-resolver
```
Comment (leeavital) [inline pkg/process/checks/net.go:176]:
```suggestion
		// Sort connections by remote IP/PID for more efficient resolution
```

Comment (ISauve) [reply pkg/process/checks/net.go]: Good catch 😬
